pdchaosazure/machine/actions.py

```diff
@@ -35,17 +36,17 @@ def delete_machines(filter: str = None,
     machines = fetch_machines(filter, configuration, secrets)
     client = init_client(secrets, configuration)
     machine_records = Records()
-    for machine in machines:
-        logger.debug("Deleting machine: {}".format(machine['name']))
 
-        try:
-            poller = client.virtual_machines.delete(machine['resourceGroup'], machine['name'])
+    with concurrent.futures.ThreadPoolExecutor(max_workers=len(machines)) as executor:
+        for machine in machines:
+            logger.debug("Deleting machine: {}".format(machine['name']))
 
-        except azure_exceptions.CloudError as e:
-            raise FailedActivity(e.message)
+            try:
+                poller = client.virtual_machines.delete(machine['resourceGroup'], machine['name'])
+            except azure_exceptions.CloudError as e:
+                raise FailedActivity(e.message)
 
-        poller.result(config.load_timeout(configuration))
-        machine_records.add(cleanse.machine(machine))
+            executor.submit(__long_poll, machine, poller, machine_records, config)
 
     return machine_records.output_as_dict('resources')
```

Comment:
Few things bother me he here. 

1. We are not waiting for results. "submit" function of ThreadPoolExecutor object is returning a "feature" on which we should call wait. 
2. We are modyfing shared object  "machine_records" in different threads - this could be unsafe and you can get strange results.

My proposoal here:
- Wait for results
- __long_poll should return "machine_records" (do not pass shared object to the thread/process)
- wait for all executors to finish and create one big machine_records set from all executors results

Reply:
First when touching thus the same came up to my mind. Then I discovered that the List object's append function is thread-safe and mutually exlusive. So we have here no problem regarding the thread-safety.

Waiting for results - the same I was thinking about as well. Buut we do not collect this kind of information and we do not process it. So that's why I ignored it. When needed the implementation is straight forward to do so.

Are there other reasons to implement it the way you proposed? Let me know

Reply:
What about machine_records. We are passing it to the executor. So we need to be sure in the return statement (line:51) that all executors finished, right? Or somehow "with" statement ensures that all executors are finished before leaving it's scope

Reply:
Exact. The with statement has a builtin .join() operator. The main thread waits until all threads joined the main thread.